js/sites.js

```diff
@@ -102,12 +102,15 @@
      * when done.
      */
     block() {
-      if (!this.isBlocked()) {
-        gUndoDialog.show(this);
-        gNewTab.sendToBrowser("NewTab:BlockLink", {
-          link: this._link
-        });
-      }
+      let self = this;
+      return async(function* () {
+        let blocked = yield self.isBlocked();
+        if (!blocked) {
+          gUndoDialog.show(self);
+          gBlockedLinks.block(self._link);
+          gNewTab.sendToBrowser("NewTab:UpdateGrid");
+        }
+      })();
```

Comment:
you probably don't need "self" for the closure, as you can call asycn with a second argument, which is `this`:

```JS
async(function* () {...}, this);
```